src/SonarScanner.MSBuild.Common/CommandLine/CommandLineParser.cs

```diff
@@ -123,7 +123,7 @@ public bool ParseArguments(IEnumerable<string> commandLineArgs, ILogger logger,
             // as possible about the failures.
             parsedOk &= CheckRequiredArgumentsSupplied(recognized, logger);
 
-            argumentInstances = parsedOk ? recognized : Enumerable.Empty<ArgumentInstance>();
+            argumentInstances = recognized;
 
             return parsedOk;
```

Comment:
What about removing line 124 and
```suggestion
            return parsedOk && CheckRequiredArgumentsSupplied(recognized, logger);
```

Reply:
I first made the change, but then I backtracked for three reasons:
# 1. Code symmetry and principle of least astonishment
Probably the weakest argument: all methods in the `ArgumentProcessor` and `CommandLineParser` follow this pattern. The result of the `bool TrySomething(..., out result)` method is always kept in a `parsedOk` and return at the end with a `return parsedOk`. Changing that in a specific method would violate POLA.

# 2. Invariant on result variable at method end
All the `bool TrySomething(..., out result)` methods need to return two things: whether they succeeded, and what's their result value. Because of the limitations we currently have with the use of `ValueTuple` and destructuring, we can't return those two things homogeneously, and we have to pass references by reference instead, that is a workaround.
To keep the code easy to reason about, seems a good norm to only assign the `out` variable at the end, and only after it was conclusively determined that the method *can* return a value.
That is done in the current version of the code: assigning `argumentInstances` is done only after having determined the value of `parsedOk` conclusively.
See point 3 for a scenario when that may not be the case, after the proposed refactoring.

# 3. Flow on exception
With this change `argumentInstances` would be assigned `recognized` even when `CheckRequiredArgumentsSupplied` raises an exception. That is a change in behavior.

I am open for discussion.